File: mobile/lib/core/models/terrain.dart
import 'package:equatable/equatable.dart';

class TerrainLayer extends Equatable {
  final String id;
  final String projectId;
  final String name;
  final String layerType;
  final String? sourceFile;
  final TerrainBounds? bounds;
  final double? resolutionM;
  final String? crs;
  final double? minElevation;
  final double? maxElevation;
  final DateTime createdAt;

  const TerrainLayer({
    required this.id,
    required this.projectId,
    required this.name,
    required this.layerType,
    this.sourceFile,
    this.bounds,
    this.resolutionM,
    this.crs,
    this.minElevation,
    this.maxElevation,
    required this.createdAt,
  });

  factory TerrainLayer.fromJson(Map<String, dynamic> json) {
    return TerrainLayer(
      id: json['id'] as String,
      projectId: json['project_id'] as String,
      name: json['name'] as String,
      layerType: json['layer_type'] as String,
      sourceFile: json['source_file'] as String?,
      bounds: json['bounds'] != null
          ? TerrainBounds.fromJson(json['bounds'] as Map<String, dynamic>)
          : null,
      resolutionM: (json['resolution_m'] as num?)?.toDouble(),
      crs: json['crs'] as String?,
      minElevation: (json['min_elevation'] as num?)?.toDouble(),
      maxElevation: (json['max_elevation'] as num?)?.toDouble(),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  @override
  List<Object?> get props => [id, projectId, layerType];
}

class TerrainBounds extends Equatable {
  final double minX;
  final double minY;
  final double maxX;
  final double maxY;

  const TerrainBounds({
    required this.minX,
    required this.minY,
    required this.maxX,
    required this.maxY,
  });

  factory TerrainBounds.fromJson(Map<String, dynamic> json) {
    return TerrainBounds(
      minX: (json['min_x'] as num).toDouble(),
      minY: (json['min_y'] as num).toDouble(),
      maxX: (json['max_x'] as num).toDouble(),
      maxY: (json['max_y'] as num).toDouble(),
    );
  }

  @override
  List<Object?> get props => [minX, minY, maxX, maxY];
}

class ElevationGrid extends Equatable {
  final int width;
  final int height;
  final List<double> elevations;
  final double minElevation;
  final double maxElevation;

  const ElevationGrid({
    required this.width,
    required this.height,
    required this.elevations,
    required this.minElevation,
    required this.maxElevation,
  });

  factory ElevationGrid.fromJson(Map<String, dynamic> json) {
    return ElevationGrid(
      width: json['width'] as int,
      height: json['height'] as int,
      elevations: (json['elevations'] as List<dynamic>)
          .map((e) => (e as num).toDouble())
          .toList(),
      minElevation: (json['min_elevation'] as num).toDouble(),
      maxElevation: (json['max_elevation'] as num).toDouble(),
    );
  }

  @override
  List<Object?> get props => [width, height, minElevation, maxElevation];
}

File: services/compute-service/internal/mappers/terrain.go
package mappers

import (
	"p9e.in/samavaya/solar3d/compute-service/internal/models"

	terrainv1 "p9e.in/samavaya/solar3d/gen/terrain/v1"
)

func ProtoToGetElevation(req *terrainv1.GetElevationRequest) *models.GetElevationRequest {
	if req == nil {
		return nil
	}
	return &models.GetElevationRequest{
		ProjectID: req.ProjectId,
		Longitude: req.Longitude,
		Latitude:  req.Latitude,
	}
}

func GetElevationToProto(resp *models.GetElevationResponse) *terrainv1.GetElevationResponse {
	if resp == nil {
		return &terrainv1.GetElevationResponse{}
	}
	return &terrainv1.GetElevationResponse{Elevation: resp.Elevation}
}

func ProtoToGetElevationGrid(req *terrainv1.GetElevationGridRequest) *models.GetElevationGridRequest {
	if req == nil || req.Bounds == nil {
		return nil
	}
	return &models.GetElevationGridRequest{
		ProjectID: req.ProjectId,
		Bounds: models.BoundingBoxModel{
			MinX: req.Bounds.MinX,
			MinY: req.Bounds.MinY,
			MaxX: req.Bounds.MaxX,
			MaxY: req.Bounds.MaxY,
		},
		ResolutionM: req.ResolutionM,
	}
}

func GetElevationGridToProto(resp *models.GetElevationGridResponse) *terrainv1.GetElevationGridResponse {
	if resp == nil {
		return &terrainv1.GetElevationGridResponse{}
	}
	return &terrainv1.GetElevationGridResponse{
		Width:        resp.Width,
		Height:       resp.Height,
		Elevations:   append([]float64(nil), resp.Elevations...),
		MinElevation: resp.MinElevation,
		MaxElevation: resp.MaxElevation,
	}
}

func ProtoToComputeSlope(req *terrainv1.ComputeSlopeRequest) *models.ComputeSlopeRequest {
	if req == nil {
		return nil
	}
	return &models.ComputeSlopeRequest{TerrainLayerID: req.TerrainLayerId}
}

func ComputeSlopeToProto(resp *models.ComputeSlopeResponse) *terrainv1.ComputeSlopeResponse {
	if resp == nil {
		return &terrainv1.ComputeSlopeResponse{}
	}
	return &terrainv1.ComputeSlopeResponse{SlopeLayer: terrainLayerToProto(resp.SlopeLayer)}
}

func ProtoToComputeAspect(req *terrainv1.ComputeAspectRequest) *models.ComputeAspectRequest {
	if req == nil {
		return nil
	}
	return &models.ComputeAspectRequest{TerrainLayerID: req.TerrainLayerId}
}

func ComputeAspectToProto(resp *models.ComputeAspectResponse) *terrainv1.ComputeAspectResponse {
	if resp == nil {
		return &terrainv1.ComputeAspectResponse{}
	}
	return &terrainv1.ComputeAspectResponse{AspectLayer: terrainLayerToProto(resp.AspectLayer)}
}

func terrainLayerToProto(in models.TerrainLayerModel) *terrainv1.TerrainLayer {
	return &terrainv1.TerrainLayer{
		Id:         in.ID,
		Name:       in.Name,
		SourceFile: in.SourceFile,
		Bounds: &terrainv1.BoundingBox{
			MinX: in.Bounds.MinX,
			MinY: in.Bounds.MinY,
			MaxX: in.Bounds.MaxX,
			MaxY: in.Bounds.MaxY,
		},
		ResolutionM:  in.ResolutionM,
		Crs:          in.CRS,
		MinElevation: in.MinElevation,
		MaxElevation: in.MaxElevation,
		LayerType:    layerTypeToProto(in.LayerType),
	}
}

func layerTypeToProto(kind string) terrainv1.TerrainLayerType {
	switch kind {
	case "DEM":
		return terrainv1.TerrainLayerType_TERRAIN_LAYER_TYPE_DEM
	case "SLOPE":
		return terrainv1.TerrainLayerType_TERRAIN_LAYER_TYPE_SLOPE
	case "ASPECT":
		return terrainv1.TerrainLayerType_TERRAIN_LAYER_TYPE_ASPECT
	case "HILLSHADE":
		return terrainv1.TerrainLayerType_TERRAIN_LAYER_TYPE_HILLSHADE
	default:
		return terrainv1.TerrainLayerType_TERRAIN_LAYER_TYPE_UNSPECIFIED
	}
}

